src/main/rx/srxl2.c

```diff
@@ -0,0 +1,610 @@
+#define SRXL2_PORT_MODE                MODE_RXTX
+
+#define SRXL2_REPLY_QUIESCENCE         (2 * 10 * 1000000 / SRXL2_PORT_BAUDRATE_DEFAULT) /*2 * (lastIdleTimestamp - lastReceiveTimestamp)*/
+
+#define SRXL2_ID                       0xA6
+#define SRXL2_MAX_PACKET_LENGTH        80
+#define SRXL2_DEVICE_ID_BROADCAST      0xFF
+
+#define SRXL2_FRAME_TIMEOUT_US         50000
+
+#define SRXL2_LISTEN_FOR_ACTIVITY_TIMEOUT_US 50000
+#define SRXL2_SEND_HANDSHAKE_TIMEOUT_US 50000
+#define SRXL2_LISTEN_FOR_HANDSHAKE_TIMEOUT_US 200000
+
+
+static uint8_t unitId = 0;
+static uint8_t baudRate = 0;
+
+static Srxl2State state = Disabled;
+static volatile uint32_t timeoutTimestamp = 0;
+static volatile uint32_t fullTimeoutTimestamp = 0;
+static volatile uint32_t lastReceiveTimestamp = 0;
+static volatile uint32_t lastValidPacketTimestamp = 0;
+static volatile uint32_t lastIdleTimestamp = 0;
+
+static volatile uint8_t readBuffer[2 * SRXL2_MAX_PACKET_LENGTH];
+static volatile uint8_t readBufferIdx = 0;
+static volatile uint8_t readBufferReadIdx = 0;
+static uint8_t writeBuffer[SRXL2_MAX_PACKET_LENGTH];
+static uint8_t writeBufferIdx = 0;
+
+static serialPort_t *serialPort;
+
+static uint8_t busMasterDeviceId = 0xFF;
+static bool telemetryRequested = false;
+
+static uint8_t telemetryFrame[22];
+
+uint8_t globalResult = 0;
+bool srxl2ProcessHandshake(const Srxl2Header* header, const Srxl2HandshakeSubHeader* handshake)
+{
+    if (handshake->destinationDeviceId == Broadcast) {
+        DEBUG("broadcast handshake from %x\r\n", handshake->sourceDeviceId);
+        busMasterDeviceId = handshake->sourceDeviceId;
+
+        if (handshake->baudSupported == 1) {
+            serialSetBaudRate(serialPort, SRXL2_PORT_BAUDRATE_HIGH);
+            DEBUG("switching to %d baud\r\n", SRXL2_PORT_BAUDRATE_HIGH);
+        }
+
+        state = Running;
+
+        return true;
+    }
+
+
+    if (handshake->destinationDeviceId != ((FlightController << 4) | unitId)) {
+        return true;
+    }
+
+    DEBUG("FC handshake from %x\r\n", handshake->sourceDeviceId);
+
+    Srxl2HandshakeFrame response = {
+        .header = *header,
+        .payload = {
+            handshake->destinationDeviceId,
+            handshake->sourceDeviceId,
+            /* priority */ 10,
+            /* baudSupported*/ baudRate,
+            /* info */ 0,
+            U_ID_2
+        }
+    };
+
+    const uint16_t crc = crc16_ccitt_update(0, &response, sizeof(response) - 2);
+    response.crcHigh = (crc >> 8);
+    response.crcLow = (crc & 0xFF);
+
+    srxl2RxWriteData(&response, sizeof(response));
+
+    return true;
+}
+
+void srxl2ProcessChannelData(const Srxl2ChannelDataHeader* channelData, rxRuntimeConfig_t *rxRuntimeConfig) {
+    if (channelData->rssi >= 0) {
+        const int rssiPercent = channelData->rssi;
+        setRssi(scaleRange(rssiPercent, 0, 100, 0, RSSI_MAX_VALUE), RSSI_SOURCE_RX_PROTOCOL);
+    } else {
+        const int rssiDbm = channelData->rssi;
+        (void) rssiDbm;
```

Comment:
What is this for?

Reply:
The protocol can provide either a positive percent value for rssi (receiver handles converting dBm to %), or a negative dBm value. This else statement appears to have been a placeholder for how to handle the negative dBm value. Not sure if dBm is something betaflight tends to handle as it is usually displayed in %.  I would add in a default conversion table, but not all receivers have the same sensitivity range, so I think I'd prefer to remove the first 2 lines, and just have it set the rssi to 50% as it currently does.